src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -296,10 +296,10 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
                         // new change starts after old change, but overlaps
                         // add as much of the old change as possible and try again
                         var oldChangeLeadingInsertion = newChange.Span.Start - (oldChange.Span.Start + oldDelta);
-                        AddRange(list, new TextChangeRange(oldChange.Span, oldChangeLeadingInsertion));
-                        oldDelta = oldDelta - oldChange.Span.Length + oldChangeLeadingInsertion;
-                        oldChange = new TextChangeRange(new TextSpan(oldChange.Span.Start, 0), oldChange.NewLength - oldChangeLeadingInsertion);
-                        newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newChange.Span.Length), newChange.NewLength);
+                        var oldChangeLeadingDeletion = Math.Min(oldChange.Span.Length, oldChangeLeadingInsertion);
```

Comment:
OK, I'm talking myself through why this makes sense.

In this line we are figuring out how many characters of the old change occur before the new change. We assume that if the old change deletes characters then it will delete from the left *at most* as many as it is inserting on the left. When the old change's deletion is longer than the old change's leading insertion, the rest of its deletion has to be shifted over to remove the characters after the new change's insertion.